app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt

```diff
@@ -0,0 +1,16 @@
+package compiler.intermediate_form
+
+import compiler.ast.Function
+import compiler.ast.Program
+import compiler.ast.Variable
+import compiler.common.semantic_analysis.ReferenceMap
+
+object FunctionDependenciesAnalyzer {
+    fun variablesUsedByFunctions(ast: Program): ReferenceMap<Function, Pair<List<Variable>, List<Variable>>> { // (accessed, modified)
+        return TODO()
+    }
+
+    fun createCallGraph(ast: Program): Set<Pair<Function, Function>> {
```

Comment:
A problem here is that we need to compare `Function`s by reference. This should be something like `ReferenceMap<Function, ReferenceSet<Function>>`.

Reply:
A `ReferenceSet` can be made in this way:
```kotlin
interface ReferenceSet<E> : Set<E>

interface MutableReferenceSet<E> : MutableSet<E>, ReferenceSet<E>

class ReferenceHashSet<E> : MutableSet<E> by Collections.newSetFromMap(ReferenceHashMap()), MutableReferenceSet<E>
```